src/System.Net.Primitives/src/System/Net/IPAddressParser.cs

```diff
@@ -66,38 +47,223 @@ internal static unsafe IPAddress Parse(string ipString, bool tryParse)
+        {
+            Debug.Assert(address != null);
+            Debug.Assert(address.Length == IPAddressParserStatics.IPv6AddressShorts);
+
+            const int INET6_ADDRSTRLEN = 65;
+            StringBuilder buffer = StringBuilderCache.Acquire(INET6_ADDRSTRLEN);
+
+            if (IPv6AddressHelper.ShouldHaveIpv4Embedded(address))
             {
-                return sb.ToString();
+                // We need to treat the last 2 ushorts as a 4-byte IPv4 address,
+                // so output the first 6 ushorts normally, followed by the IPv4 address.
+                AppendSections(address, 0, 6, buffer);
+                if (buffer[buffer.Length - 1] != ':')
+                {
+                    buffer.Append(':');
+                }
+                buffer.Append(IPAddressParser.IPv4AddressToString(ExtractIPv4Address(address)));
             }
             else
             {
-                throw new SocketException(IPAddressPal.GetSocketErrorForErrorCode(errorCode), errorCode);
+                // No IPv4 address.  Output all 8 sections as part of the IPv6 address
+                // with normal formatting rules.
+                AppendSections(address, 0, 8, buffer);
+            }
+
+            // If there's a scope ID, append it.
+            if (scopeId != 0)
+            {
+                buffer.Append('%').Append(scopeId);
             }
+
+            return StringBuilderCache.GetStringAndRelease(buffer);
         }
 
-        internal static string IPv6AddressToString(byte[] numbers, uint scopeId)
+        private static unsafe void FormatIPv4AddressNumber(int number, char* addressString, ref int offset)
         {
-            StringBuilder sb = new StringBuilder(INET6_ADDRSTRLEN);
-            uint errorCode = IPAddressPal.Ipv6AddressToString(numbers, scopeId, sb);
+            int i = offset;
+            do
+            {
+                int rem;
+                number = Math.DivRem(number, 10, out rem);
+                addressString[--i] = (char)('0' + rem);
+            } while (number != 0);
+            offset = i;
+        }
+
+        public static unsafe bool Ipv4StringToAddress(string ipString, out long address)
+        {
+            Debug.Assert(ipString != null);
+
+            long tmpAddr;
+            int end = ipString.Length;
+            fixed (char* ipStringPtr = ipString)
+            {
+                tmpAddr = IPv4AddressHelper.ParseNonCanonical(ipStringPtr, 0, ref end, notImplicitFile: true);
+            }
 
-            if (errorCode == IPAddressPal.SuccessErrorCode)
+            if (tmpAddr != IPv4AddressHelper.Invalid && end == ipString.Length)
             {
-                return sb.ToString();
+                address =
+                    ((0xFF000000 & tmpAddr) >> 24) |
+                    ((0x00FF0000 & tmpAddr) >> 8) |
+                    ((0x0000FF00 & tmpAddr) << 8) |
+                    ((0x000000FF & tmpAddr) << 24);
+                return true;
```

Comment:
Can we at least add a comment here?  I'm not sure why IPv4AddressHelper returns the bytes in the reverse order.